src/libraries/System.Net.Quic/src/System/Net/Quic/QuicListener.PendingConnection.cs

```diff
@@ -0,0 +1,61 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Net.Security;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Net.Quic;
+
+public sealed partial class QuicListener
+{
+    private sealed class PendingConnection : IAsyncDisposable
+    {
+        private static readonly TimeSpan s_handshakeTimeout = TimeSpan.FromSeconds(10);
+
+        private readonly TaskCompletionSource<QuicConnection?> _finishHandshakeTask;
+        private readonly CancellationTokenSource _cancellationTokenSource;
+
+        public PendingConnection()
+        {
+            _finishHandshakeTask = new TaskCompletionSource<QuicConnection?>();
```

Comment:
Should this use RunContinuationsAsynchronously?

Reply:
Yes.